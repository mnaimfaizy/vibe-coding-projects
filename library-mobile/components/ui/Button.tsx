/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';

import { StyleSheet, View } from 'react-native';

import { Button as PaperButton } from 'react-native-paper';

interface ButtonProps {
  title: string;
  variant?: 'primary' | 'secondary' | 'outline' | 'text';
  loading?: boolean;
  fullWidth?: boolean;
  onPress?: () => void;
  disabled?: boolean;
  style?: any;
}

export const Button: React.FC<ButtonProps> = ({
  title,
  variant = 'primary',
  loading = false,
  fullWidth = true,
  disabled,
  style,
  onPress,
  ...props
}) => {
  const getButtonMode = (): 'text' | 'outlined' | 'contained' | 'elevated' | 'contained-tonal' => {
    switch (variant) {
      case 'primary':
        return 'contained';
      case 'secondary':
        return 'contained-tonal';
      case 'outline':
        return 'outlined';
      case 'text':
        return 'text';
      default:
        return 'contained';
    }
  };

  return (
    <View style={[fullWidth && styles.fullWidth, style]}>
      <PaperButton
        mode={getButtonMode()}
        onPress={onPress}
        loading={loading}
        disabled={disabled || loading}
        style={styles.button}
        contentStyle={styles.buttonContent}
        labelStyle={styles.buttonLabel}>
        {title}
      </PaperButton>
    </View>
  );
};

const styles = StyleSheet.create({
  fullWidth: {
    width: '100%',
  },
  button: {
    borderRadius: 8,
  },
  buttonContent: {
    height: 48,
  },
  buttonLabel: {
    fontSize: 16,
    fontWeight: '600',
  },
});
