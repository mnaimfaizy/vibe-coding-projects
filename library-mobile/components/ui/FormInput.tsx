import React, { useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { HelperText, TextInput, useTheme } from 'react-native-paper';

interface FormInputProps extends React.ComponentProps<typeof TextInput> {
  label: string;
  error?: string;
  secureTextEntry?: boolean;
  icon?: string;
}

export const FormInput: React.FC<FormInputProps> = ({
  label,
  error,
  secureTextEntry,
  icon,
  ...props
}) => {
  const [isFocused, setIsFocused] = useState(false);
  const [showPassword, setShowPassword] = useState(false);

  const { colors } = useTheme();

  return (
    <View style={styles.container}>
      <TextInput
        label={label}
        secureTextEntry={secureTextEntry && !showPassword}
        onFocus={() => setIsFocused(true)}
        onBlur={() => setIsFocused(false)}
        mode="outlined"
        error={!!error}
        left={icon ? <TextInput.Icon icon={icon} /> : undefined}
        right={
          secureTextEntry ? (
            <TextInput.Icon
              icon={showPassword ? 'eye-off' : 'eye'}
              onPress={() => setShowPassword(!showPassword)}
            />
          ) : undefined
        }
        style={styles.input}
        outlineStyle={styles.outline}
        {...props}
      />
      {error ? (
        <HelperText type="error" visible={!!error} style={styles.error}>
          {error}
        </HelperText>
      ) : null}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
    width: '100%',
  },
  input: {
    backgroundColor: 'transparent',
  },
  outline: {
    borderRadius: 8,
  },
  error: {
    marginBottom: 0,
    paddingHorizontal: 0,
  },
});
