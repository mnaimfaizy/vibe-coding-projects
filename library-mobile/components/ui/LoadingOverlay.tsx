import React from 'react';
import { StyleSheet, View } from 'react-native';
import { ActivityIndicator, Surface, Text, useTheme } from 'react-native-paper';
import { useThemeColor } from '../../hooks/useThemeColor';

interface LoadingOverlayProps {
  message?: string;
}

export const LoadingOverlay: React.FC<LoadingOverlayProps> = ({ message = 'Loading...' }) => {
  const backgroundColor = useThemeColor({}, 'background');
  const { colors } = useTheme();
  
  return (
    <View style={[styles.container, { backgroundColor: backgroundColor + 'dd' }]}>
      <Surface style={styles.loadingContainer} elevation={4}>
        <ActivityIndicator size="large" color={colors.primary} animating={true} />
        <Text variant="bodyLarge" style={styles.message}>{message}</Text>
      </Surface>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    ...StyleSheet.absoluteFillObject,
    justifyContent: 'center',
    alignItems: 'center',
    zIndex: 1000,
  },
  loadingContainer: {
    padding: 24,
    borderRadius: 12,
    alignItems: 'center',
    justifyContent: 'center',
    minWidth: 200,
  },
  message: {
    marginTop: 16,
    fontWeight: '500',
  },
});