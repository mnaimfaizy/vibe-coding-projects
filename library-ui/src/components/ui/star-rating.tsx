// filepath: c:\Users\mnaim\Downloads\Projects\vibe-coding-projects\library-ui\src\components\ui\star-rating.tsx
import React from "react";
import { Star } from "lucide-react";
import { cn } from "@/lib/utils";

interface StarRatingProps {
  rating: number;
  maxRating?: number;
  size?: "sm" | "md" | "lg";
  interactive?: boolean;
  onRatingChange?: (newRating: number) => void;
  className?: string;
}

export function StarRating({
  rating,
  maxRating = 5,
  size = "md",
  interactive = false,
  onRatingChange,
  className,
}: StarRatingProps) {
  const [hoverRating, setHoverRating] = React.useState(0);

  const sizes = {
    sm: "h-4 w-4",
    md: "h-5 w-5",
    lg: "h-6 w-6",
  };

  const sizeClass = sizes[size];

  return (
    <div
      className={cn("flex items-center", className)}
      onMouseLeave={() => interactive && setHoverRating(0)}
    >
      {Array.from({ length: maxRating }).map((_, i) => {
        const starValue = i + 1;
        const isActive = interactive
          ? starValue <= (hoverRating || rating)
          : starValue <= rating;

        return (
          <Star
            key={i}
            className={cn(
              sizeClass,
              "transition-all duration-100",
              isActive ? "text-yellow-400 fill-yellow-400" : "text-gray-300",
              interactive && "cursor-pointer hover:scale-110"
            )}
            onClick={() => {
              if (interactive && onRatingChange) {
                // Toggle off if clicking the same star
                onRatingChange(
                  rating === starValue ? starValue - 1 : starValue
                );
              }
            }}
            onMouseEnter={() => interactive && setHoverRating(starValue)}
          />
        );
      })}

      {/* Display numeric rating if not interactive */}
      {!interactive && (
        <span className="ml-2 text-sm text-gray-600 dark:text-gray-400">
          {rating}/{maxRating}
        </span>
      )}
    </div>
  );
}
