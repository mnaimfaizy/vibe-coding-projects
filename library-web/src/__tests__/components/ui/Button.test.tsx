import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { Button } from "../../../components/ui/button";

describe("Button Component", () => {
  it("renders default button correctly", () => {
    const { container } = render(<Button>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  it("renders different variants correctly", () => {
    const variants = [
      "default",
      "destructive",
      "outline",
      "secondary",
      "ghost",
      "link",
    ] as const;

    variants.forEach((variant) => {
      const { container } = render(
        <Button variant={variant}>Button {variant}</Button>
      );
      expect(container).toMatchSnapshot();
    });
  });

  it("renders different sizes correctly", () => {
    const sizes = ["default", "sm", "lg", "icon"] as const;

    sizes.forEach((size) => {
      const { container } = render(<Button size={size}>Button {size}</Button>);
      expect(container).toMatchSnapshot();
    });
  });

  it("applies custom className", () => {
    const { container } = render(
      <Button className="custom-class">Custom Button</Button>
    );
    expect(container).toMatchSnapshot();
    expect(container.firstChild).toHaveClass("custom-class");
  });

  it("renders as a link when asChild is used with anchor", () => {
    const { container } = render(
      <Button asChild>
        <a href="https://example.com">Link Button</a>
      </Button>
    );
    expect(container).toMatchSnapshot();
    expect(container.querySelector("a")).toHaveAttribute(
      "href",
      "https://example.com"
    );
  });

  it("forwards props correctly", () => {
    const { container } = render(
      <Button disabled aria-label="Test Button" data-testid="test-button">
        Disabled Button
      </Button>
    );
    expect(container).toMatchSnapshot();
    expect(container.firstChild).toHaveAttribute("disabled", "");
    expect(container.firstChild).toHaveAttribute("aria-label", "Test Button");
    expect(container.firstChild).toHaveAttribute("data-testid", "test-button");
  });
});
