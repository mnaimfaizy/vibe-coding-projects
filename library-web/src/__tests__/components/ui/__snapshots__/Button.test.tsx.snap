// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > applies custom className 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 custom-class"
    data-slot="button"
  >
    Custom Button
  </button>
</div>
`;

exports[`Button Component > forwards props correctly 1`] = `
<div>
  <button
    aria-label="Test Button"
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
    data-testid="test-button"
    disabled=""
  >
    Disabled Button
  </button>
</div>
`;

exports[`Button Component > renders as a link when asChild is used with anchor 1`] = `
<div>
  <a
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
    href="https://example.com"
  >
    Link Button
  </a>
</div>
`;

exports[`Button Component > renders default button correctly 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Click me
  </button>
</div>
`;

exports[`Button Component > renders different sizes correctly 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Button 
    default
  </button>
</div>
`;

exports[`Button Component > renders different sizes correctly 2`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
    data-slot="button"
  >
    Button 
    sm
  </button>
</div>
`;

exports[`Button Component > renders different sizes correctly 3`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-10 rounded-md px-6 has-[>svg]:px-4"
    data-slot="button"
  >
    Button 
    lg
  </button>
</div>
`;

exports[`Button Component > renders different sizes correctly 4`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 size-9"
    data-slot="button"
  >
    Button 
    icon
  </button>
</div>
`;

exports[`Button Component > renders different variants correctly 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Button 
    default
  </button>
</div>
`;

exports[`Button Component > renders different variants correctly 2`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-destructive text-white shadow-xs hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Button 
    destructive
  </button>
</div>
`;

exports[`Button Component > renders different variants correctly 3`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Button 
    outline
  </button>
</div>
`;

exports[`Button Component > renders different variants correctly 4`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-secondary text-secondary-foreground shadow-xs hover:bg-secondary/80 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Button 
    secondary
  </button>
</div>
`;

exports[`Button Component > renders different variants correctly 5`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Button 
    ghost
  </button>
</div>
`;

exports[`Button Component > renders different variants correctly 6`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer text-primary underline-offset-4 hover:underline h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Button 
    link
  </button>
</div>
`;
